test/p_ossltest.c

```diff
@@ -692,6 +693,345 @@ static const OSSL_DISPATCH ossl_testaes128_cbc_functions[] = {
+     */
+    inbuf = OPENSSL_zalloc(inl);
+    if (inbuf == NULL)
+        return 0;
+
+    memcpy(inbuf, in, inl);
+
+    soutl = EVP_Cipher(ctx->sub_ctx, (unsigned char *)out, in, (unsigned int)inl);
+
+    if (soutl <= 0)
+        goto err;
+
+    /*
+     * replace the ciphertext with our plain text
+     */
+    memcpy(out, inbuf, inl);
+
+    ret = 1;
+    *outl = soutl;
+    err:
+    OPENSSL_free(inbuf);
+    return ret;
+}
+
+/**
+ * @brief Finalize the operation and produce any remaining output for aes-ecb-128
+ *
+ * @param vprovctx void *vprovctx.
+ * @param out unsigned char *out.
+ * @param outl size_t *outl.
+ * @param outsize size_t outsize.
+ * @return int.
+ */
+
+static int ossl_test_aes128ecb_final(void *vprovctx, unsigned char *out, size_t *outl,
+                                     size_t outsize)
+{
+    PROV_EVP_AES128_ECB_CTX *ctx = (PROV_EVP_AES128_ECB_CTX *)vprovctx;
+    int soutl;
+    int ret;
+
+    ret = EVP_CipherFinal_ex(ctx->sub_ctx, out, &soutl);
+
+    return ret;
+}
+
+/**
+ * @brief Implement ossl test aes128ecb cipher.
+ *
+ * Note, nothing in TLS should be using this function, as we are a provider
+ * we just need it for completeness.
+ *
+ * @param vprovctx void *vprovctx.
+ * @param out unsigned char *out.
+ * @param outl size_t *outl.
+ * @param outsize size_t outsize.
+ * @param in const unsigned char *in.
+ * @param inl size_t inl.
+ * @return int.
+ */
+
+static int ossl_test_aes128ecb_cipher(void *vprovctx, unsigned char *out, size_t *outl,
+                                      size_t outsize, const unsigned char *in, size_t inl)
+{
+    PROV_EVP_AES128_ECB_CTX *ctx = (PROV_EVP_AES128_ECB_CTX *)vprovctx;
+
+    return EVP_Cipher(ctx->sub_ctx, out, in, (int) inl);
+}
+
+/**
+ * @brief Return provider or algorithm parameters.
+ *
+ * @param params OSSL_PARAM params[].
+ * @return int.
+ */
+
+static int ossl_test_aes128ecb_get_params(OSSL_PARAM params[])
+{
+    return ossl_cipher_generic_get_params(params, EVP_CIPH_ECB_MODE, 0, 128, 128, 0);
+}
+
+/**
+ * @brief Query parameters from the aes-128-ecb context.
+ *
+ * @param vprovctx void *vprovctx.
+ * @param params OSSL_PARAM params[].
+ * @return int.
+ */
+
+static int ossl_test_aes128ecb_get_ctx_params(void *vprovctx, OSSL_PARAM params[])
+{
+    PROV_EVP_AES128_ECB_CTX *ctx = (PROV_EVP_AES128_ECB_CTX *)vprovctx;
+
+    return EVP_CIPHER_CTX_get_params(ctx->sub_ctx, params);
+}
+
+/**
+ * @brief Set parameters on the aes-128-ecb context.
+ *
+ * @param vprovctx void *vprovctx.
+ * @param params const OSSL_PARAM params[].
+ * @return int.
+ */
+
+static int ossl_test_aes128ecb_set_ctx_params(void *vprovctx, const OSSL_PARAM params[])
+{
+    /*
+     * Normally this gets called to set
+     * OSSL_CIPHER_PARAM_TLS_VERSION
+     * OSSL_CIPHER_PARAM_TLS_MAC_SIZE
+     * but we don't want the underlying cipher to do that, we will
+     * handle that padding in ecb_update on our own, so intercept and
+     * squash that here
+     */
+    return 1;
```

Comment:
This needs to be properly wired up for this case. Unlike aes-128-cbc this is not being used for a TLS ciphersuite. This is just being used a standard EVP cipher. in particular we need to pass down the parameters to turn off padding.


```suggestion
    PROV_EVP_AES128_ECB_CTX *ctx = (PROV_EVP_AES128_ECB_CTX *)vprovctx;

    return EVP_CIPHER_CTX_set_params(ctx->sub_ctx, params);
```